presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -181,7 +182,30 @@ public GroupedAccumulator createGroupedAccumulator(UpdateMemory updateMemory)
         ImmutableList.Builder<Integer> aggregateInputChannels = ImmutableList.builder();
         aggregateInputChannels.addAll(inputChannels);
         maskChannel.ifPresent(aggregateInputChannels::add);
-        return new SpillableFinalOnlyGroupedAccumulator(sourceTypes, aggregateInputChannels.build(), (FinalOnlyGroupedAccumulator) accumulator);
+
+        Optional<MarkDistinctHash> markDistinctHash = Optional.empty();
+        if (hasDistinct()) {
+            ImmutableList.Builder<Type> inputTypes = ImmutableList.builder();
+            inputTypes.add(BIGINT); // group id column
+            inputChannels.forEach(channel -> inputTypes.add(sourceTypes.get(channel)));
+
+            IntStream.Builder inputs = IntStream.builder();
+            inputs.accept(0);  // First channel is for group id column
+            inputChannels.forEach(channel -> inputs.accept(channel + 1));
+
+            markDistinctHash = Optional.of(
+                    new MarkDistinctHash(
+                            session,
+                            inputTypes.build(),
+                            inputs.build().toArray(),
+                            Optional.empty(),
+                            joinCompiler,
+                            () -> {
+                                updateMemory.update();
```

Comment:
Preserve the comments here? Or maybe don't copy paste it and make it a shared lamba in all 3 instances in the file?